lib/datadog/core/configuration/option.rb

```diff
@@ -38,22 +58,55 @@ def initialize(definition, context)
         # @param value [Object] the new value to be associated with this option
         # @param precedence [Precedence] from what precedence order this new value comes from
         def set(value, precedence: Precedence::PROGRAMMATIC)
-          # Cannot override higher precedence value
-          if precedence[0] < @precedence_set[0]
+          # Is there a higher precedence value set?
+          if @precedence_set > precedence
+            # This should be uncommon, as higher precedence values tend to
+            # happen later in the application lifecycle.
             Datadog.logger.info do
-              "Option '#{definition.name}' not changed to '#{value}' (precedence: #{precedence[1]}) because the higher " \
-                "precedence value '#{@value}' (precedence: #{@precedence_set[1]}) was already set."
+              "Option '#{definition.name}' not changed to '#{value}' (precedence: #{precedence.name}) because the higher " \
+                "precedence value '#{@value}' (precedence: #{@precedence_set.name}) was already set."
             end
 
+            # But if it happens, we have to store the lower precedence value `value`
+            # because it's possible to revert to it by `#unset`ting
+            # the existing, higher-precedence value.
+            # Effectively, we always store one value pre precedence.
+            old_value = @value_per_precedence[precedence]
+            @value_per_precedence[precedence] = context_exec(value, old_value, &definition.setter)
```

Comment:
We execute the setter block twice when setting lower precedence, once here and the other on line 98, when we unset and find the closer precedence value store and restore the option to the previous value. 

This could become an issue if the setter blocks logic runs expensive code or modify the internal state.

Reply:
Well pointed out!
Actually this would be a correctness issue, given `setter` would be run twice for the same value, probably yield unexpected results.

Reply:
Fixed it, thank you again!